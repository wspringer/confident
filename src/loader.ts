import {inspect} from './inspector';
import * as yup from 'yup';
import {ObjectShape} from 'yup/lib/object';
import {CustomError} from 'ts-custom-error';
import chalk from 'chalk';

export type RecursivePartial<T> = {
  [P in keyof T]?: RecursivePartial<T[P]>;
};

export interface Loader<T1, T2 extends yup.ObjectSchema<ObjectShape, T1>> {
  load: (defaultOptions?: RecursivePartial<yup.Asserts<T2>>) => yup.Asserts<T2>;
  loadOrFail: (
    defaultOptions?: RecursivePartial<yup.Asserts<T2>>
  ) => yup.Asserts<T2>;
  printUsage: () => void;
}

export class LoaderError extends CustomError {
  public constructor(msg: string, public envVar?: string) {
    super(msg);
  }
}

export const loaderFrom = <T1, T2 extends yup.ObjectSchema<ObjectShape, T1>>(
  schema: T2
): Loader<T1, T2> => {
  const entries = inspect(schema);

  const load = (
    defaultOptions?: RecursivePartial<yup.Asserts<T2>>
  ): yup.Asserts<T2> => {
    const options = defaultOptions || {};
    entries.forEach(entry => {
      const value = process.env[entry.envVar];
      if (value) entry.set(options, value);
    });
    try {
      return schema.validateSync(options);
    } catch (err) {
      if (err instanceof yup.ValidationError) {
        const path = err.path;
        const name = entries.find(entry => entry.path === path)?.envVar;
        if (name) {
          throw new LoaderError(err.message, name);
        } else {
          throw new LoaderError(err.message);
        }
      }
      throw err;
    }
  };

  const loadOrFail = (defaultOptions?: RecursivePartial<yup.Asserts<T2>>) => {
    try {
      return load(defaultOptions);
    } catch (err) {
      if (err instanceof LoaderError) {
        console.error(
          chalk.red(`Failed to load configuration; ${err.message}`)
        );
        console.error(
          `Consider checking the value of the ${err.envVar} environment variable`
        );
      } else if (err instanceof Error) {
        console.error(chalk.red(err.message));
      }
      // eslint-disable-next-line no-process-exit
      process.exit(1);
    }
  };

  const printUsage = () => {
    const usage = entries
      .map(entry =>
        `
${chalk.blue(entry.envVar)}
${entry.desc || ''}
    `.trim()
      )
      .join('\n\n');
    console.info(usage);
  };

  return {
    load,
    loadOrFail,
    printUsage,
  };
};
