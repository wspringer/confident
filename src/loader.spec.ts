import {loaderFrom, LoaderError} from './loader';
import * as yup from 'yup';
import chalk from 'chalk';

// Forcing chalk to produce colors
chalk.level = 2;

const schema = yup.object({
  path: yup.string().required().meta({
    desc: 'The path to search for executables',
  }),
  zzz: yup.number().integer(),
});

describe('the configuration solution', () => {
  const consoleInfoSpy = jest.spyOn(console, 'info').mockImplementation();
  const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation();
  const processExitSpy = jest.spyOn(process, 'exit').mockImplementation();

  afterEach(() => {
    consoleInfoSpy.mockClear();
    consoleErrorSpy.mockClear();
    processExitSpy.mockClear();
  });

  afterAll(() => {
    consoleInfoSpy.mockRestore();
    consoleErrorSpy.mockRestore();
    processExitSpy.mockRestore();
  });

  it('should allow you grab the configuration', () => {
    const loader = loaderFrom(schema);
    const config = loader.load();
    expect(config.path).not.toEqual('');
  });

  it('should allow you to rely on defaults', () => {
    const loader = loaderFrom(schema);
    const config = loader.load({zzz: 2});
    expect(config.zzz).toEqual(2);
  });

  it('should allow you to print usage', () => {
    const loader = loaderFrom(schema);
    loader.printUsage();
    expect(consoleInfoSpy).toBeCalledTimes(1);
    expect(consoleInfoSpy.mock.calls[0][0]).toMatchSnapshot();
  });

  it('should inform you about schema violations', () => {
    const loader = loaderFrom(
      yup.object({
        zzz: yup.object({yyy: yup.number().required()}),
      })
    );
    expect(() => loader.load()).toThrow(LoaderError);
    expect(() => loader.load()).toThrow('zzz.yyy is a required field');
  });

  it('should print relevant messages', () => {
    const loader = loaderFrom(
      yup.object({
        zzz: yup.object({yyy: yup.number().required()}),
      })
    );
    loader.loadOrFail();
    expect(processExitSpy).toBeCalled();
    expect(processExitSpy.mock.calls[0][0]).toEqual(1);
    expect(consoleErrorSpy).toBeCalledTimes(2);
    expect(consoleErrorSpy.mock.calls[0][0]).toEqual(
      chalk.red('Failed to load configuration; zzz.yyy is a required field')
    );
    expect(consoleErrorSpy.mock.calls[1][0]).toEqual(
      'Consider checking the value of the ZZZ_YYY environment variable'
    );
  });
});
