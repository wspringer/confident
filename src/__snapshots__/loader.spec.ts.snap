// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the configuration solution should allow you to print usage 1`] = `
"[34mPATH[39m
The path to search for executables

[34mZZZ[39m"
`;
